homeworks/olena.ambitska_elena-ambitska/GamesProject/scripts/login.js

```diff
@@ -0,0 +1,107 @@
+const VALIDATE_RULES = {
+    email: {
+        rule: (email) => /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email),
+        message: 'Invalid email format',
+        required: true,
+    },
+    password: {
+        rule: (password) => password.length >= 8 && password.length <= 12,
+        message: 'Password must be between 8 and 12 characters',
+        required: true,
+    },
+    checkbox: {
+        rule: (checkbox) => checkbox,
+        message: 'Checkbox must be checked',
+        required: true,
+    },
+};
+
+const ERROR_STATE = Object.fromEntries(Object.keys(VALIDATE_RULES).map((key) => [key, false]));
+
+function displayErrorMessage(field, message) {
+    const errorContainer = document.createElement('div');
+    errorContainer.classList.add('error-message');
+    errorContainer.setAttribute('data-field', field);
+    errorContainer.innerText = message;
+
+    const inputField = document.getElementById(field);
+    const formGroup = inputField.closest('.form-group');
+    formGroup.appendChild(errorContainer);
+}
+
+function validateField(field) {
+    const value = document.getElementById(field).value.trim();
+    const rule = VALIDATE_RULES[field];
+    const isChecked = document.getElementById(field).checked;
+
+    const isFieldEmpty = rule.required && !value;
+    const isCheckboxInvalid = field === 'checkbox' && !rule.rule(isChecked);
+    const isValueInvalid = rule.rule && !rule.rule(value);
+
+    // eslint-disable-next-line object-curly-newline
+    return { value, rule, isChecked, isFieldEmpty, isCheckboxInvalid, isValueInvalid };
+}
+
+function clearErrorMessages(field) {
+    if (field) {
+        const errorMessages = document.querySelectorAll(`.error-message[data-field="${field}"]`);
+        errorMessages.forEach((errorMessage) => errorMessage.remove());
+        ERROR_STATE[field] = false;
+    } else {
+        const errorMessages = document.querySelectorAll('.error-message');
+        errorMessages.forEach((errorMessage) => errorMessage.remove());
+        Object.keys(VALIDATE_RULES).forEach((key) => {
+            ERROR_STATE[key] = false;
+        });
+    }
+}
+
+function validateInputField(field) {
+    clearErrorMessages(field);
+
+    // eslint-disable-next-line object-curly-newline
+    const { rule, isFieldEmpty, isCheckboxInvalid, isValueInvalid } = validateField(field);
```

Comment:
Same here. If you refactor your code in this way, you can avoid long line and make your variables more easy to read. 
```js
    const {
        rule,
        isFieldEmpty,
        isCheckboxInvalid,
        isValueInvalid,
    } = validateField(field);
```
Also, jsdoc or TypeScript could help to make your code self-documented (just a small piece of advice for the future).

Reply:
@vladyslav-yermolin-moc I`m waiting for the advanced course for TS :innocent: